### PR TITLE
#4378 Implemented Separating Axis Theorem for BoundingBox

### PR DESCRIPTION
**Pull request by trommlbomml**
Added Separating Axis Theorem for BoundingBox to BoundingBox collision. Also added doc comments for public Methods.

### CODE REVIEW

File: MonoGame.Framework/BoundingBox.cs [hunk 1]
```diff
@@ -382,11 +487,85 @@ public void Intersects(ref BoundingBox box, out bool result)
             return;
         }
 
+        /// <summary>
+        /// Checks for intersection to another bounding box and returning information about separtion.
```
Comment (danzel) [inline MonoGame.Framework/BoundingBox.cs]: "separtion" spelling

**Comment by tomspilman** [comment]
This wasn't your fault @PFriedland89.

@hach-que -  Why would we suddenly get this exception from Protobuild?

```
System.InvalidOperationException: There is an error in XML document. ---> System.InvalidOperationException: <Module xmlns=''> was not expected
  at System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot (System.Xml.Serialization.XmlTypeMapping rootMap) [0x00000] in <filename unknown>:0 
  at System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot () [0x00000] in <filename unknown>:0 
  at System.Xml.Serialization.XmlSerializer.Deserialize (System.Xml.Serialization.XmlSerializationReader reader) [0x00000] in <filename unknown>:0 
  --- End of inner exception stack trace ---
  at System.Xml.Serialization.XmlSerializer.Deserialize (System.Xml.Serialization.XmlSerializationReader reader) [0x00000] in <filename unknown>:0 
  at System.Xml.Serialization.XmlSerializer.Deserialize (System.Xml.XmlReader xmlReader) [0x00000] in <filename unknown>:0 
  at System.Xml.Serialization.XmlSerializer.Deserialize (System.IO.TextReader textReader) [0x00000] in <filename unknown>:0 
  at Protobuild.ModuleInfo.Load (System.String xmlFile) [0x00000] in <filename unknown>:0 
  at Protobuild.ModuleInfo.GetSubmodules (System.String platform) [0x00000] in <filename unknown>:0 
  at Protobuild.Tasks.GenerateProjectsTask.Execute () [0x00000] in <filename unknown>:0 
  at Protobuild.ActionDispatch.GenerateProjectsForPlatform (Protobuild.ModuleInfo module, System.String platform, System.String[] enabledServices, System.String[] disabledServices, System.String serviceSpecPath) [0x00000] in <filename unknown>:0 
  at Protobuild.ActionDispatch.PerformAction (Protobuild.ModuleInfo module, System.String action, System.String platform, System.String[] enabledServices, System.String[] disabledServices, System.String serviceSpecPath) [0x00000] in <filename unknown>:0 
  at Protobuild.GenerateCommand.Execute (Protobuild.Execution execution) [0x00000] in <filename unknown>:0 
  at Protobuild.MainClass.Main (System.String[] args) [0x00000] in <filename unknown>:0 
```

I'm kicking off the build again to see if it goes away or if it is a persistent error.

**Comment by hach-que** [comment]
The Module.xml file is incorrectly formed.

**Comment by KonajuGames** [comment]
Two new files were added in this PR but they weren't added to the
protobuild definition.​

**Comment by trommlbomml** [comment]
> This wasn't your fault @PFriedland89.

Was my wrong login. So this is also not "my" fault? Protobuild adds new files automatically to projects to build?

**Comment by hach-que** [comment]
I can't see anything in the repo or this patch that would cause this error.

What does Module.xml and Build/Projects files look like on the build server?

**Comment by hach-que** [comment]
This part:

[exec] at System.Xml.Serialization.XmlSerializer.Deserialize
(System.IO.TextReader textReader) [0x00000] in <filename unknown>:0
[10:52:44][exec] at Protobuild.ModuleInfo.Load (System.String xmlFile)
[0x00000] in <filename unknown>:0
[10:52:44][exec] at Protobuild.ModuleInfo.GetSubmodules (System.String
platform) [0x00000] in <filename unknown>:0
[10:52:44][exec] at Protobuild.Tasks.GenerateProjectsTask.Execute ()
[0x00000] in <filename unknown>:0

Implies the error happened while trying to load submodules... but neither
of the submodules in MonoGame have a Build folder or Protobuild, so there's
no reason for it to even attempt to call Load...

My guess is something is weird with the working copy on the build server,
but without access to that so I can see what the contents of the folder is,
I can't explain why the error occurred.

**Comment by KonajuGames** [comment]
If you add new files to the project, you need to add them to the definition
file for that project in the Build directory.

**Comment by trommlbomml** [comment]
> If you add new files to the project, you need to add them to the definition
file for that project in the Build directory.

Would be nice to point who has something to do.

**Comment by KonajuGames** [comment]
The person who added the new .cs files has to add them to MonoGame/Build/Projects/MonoGame.Framework.definition as well.  This is the file that is used by Protobuild to generate the csproj files.

**Comment by trommlbomml** [comment]
Great, thats an information that helps me to repair the issue. As this isn't explained in the contribution guide (I know there is an issue opened with it), It would be helpful to tell me fix it...I'll do with the next commit.

**Comment by trommlbomml** [comment]
Well doesn't work either. anything else I forgot?

**Comment by hach-que** [comment]
When was this PR branch made? Maybe you need to rebase on the latest
develop?

**Comment by cra0zy** [comment]
> When was this PR branch made? Maybe you need to rebase on the latest develop?

I think you are right, 559 commits behind: https://github.com/trommlbomml/MonoGame/tree/SAT

@trommlbomml do:
```
git add upstream https://github.com/mono/MonoGame.git
git rebase upstream/develop
git push --force
```

**Comment by KonajuGames** [comment]
Forgot the "remote" between "git" and "add".

**Comment by cra0zy** [comment]
> Forgot the "remote" between "git" and "add".

Thanks for notifying me on a type, fixed.

**Comment by trommlbomml** [comment]
Ok I did the steps provided from @cra0zy. hopefully I didn't broke any history.

**Comment by cra0zy** [comment]
@trommlbomml Just FYI, the build server is down right now for maintenance.

**Comment by trommlbomml** [comment]
how about returning of the build server? saw a twitter post about a new machine ;)

**Comment by cra0zy** [comment]
It's up, you will need to rebase your branch, do:
```
git remote add upstream https://github.com/mono/MonoGame.git
git fetch upstream
git rebase upstream/develop
git push --force
```

**Comment by trommlbomml** [comment]
I already did, do I need to repeat the process?

**Comment by cra0zy** [comment]
> I already did, do I need to repeat the process?

Yes.

**Comment by trommlbomml** [comment]
Done. Hopefully it works now. Is every contribution so complicated or did I just missed to do a lot of things?

**Comment by cra0zy** [comment]
> Done. Hopefully it works now. Is every contribution so complicated or did I just missed to do a lot of things?

You didn't do it, your branch is still 50+ commits behind: https://github.com/trommlbomml/MonoGame/tree/SAT

**Comment by trommlbomml** [comment]
Grml. Now it is 3 commits ahead.

**Comment by cra0zy** [comment]
```
[02:33:22][exec]   c:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\MonoGame.Framework\BoundingBox.cs(495,62): error CS0246: The type or namespace name 'SatIntersectionResult' could not be found (are you missing a using directive or an assembly reference?) [C:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\Tools\2MGFX\2MGFX.Windows.csproj]
[02:33:22][exec]   c:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\MonoGame.Framework\BoundingBox.cs(547,16): error CS0246: The type or namespace name 'SatIntersectionResult' could not be found (are you missing a using directive or an assembly reference?) [C:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\Tools\2MGFX\2MGFX.Windows.csproj]
[02:33:22][exec]   c:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\MonoGame.Framework\BoundingBox.cs(738,24): error CS0246: The type or namespace name 'Line1D' could not be found (are you missing a using directive or an assembly reference?) [C:\BuildAgents\MonoGameWin1\work\f26e5a04a66fcbdb\Tools\2MGFX\2MGFX.Windows.csproj]
```

**Comment by trommlbomml** [comment]
As everthing seems to work I would really appreciate to see this feature assigned to 3.6. Is that possible?

File: MonoGame.Framework/BoundingBox.cs [hunk 2]
```diff
@@ -16,12 +19,21 @@ public struct BoundingBox : IEquatable<BoundingBox>
 
         #region Public Fields
 
+        /// <summary>
+        /// Minimal edge of the box.
```
Comment (mgarstenauer) [inline MonoGame.Framework/BoundingBox.cs]: This should probably be "Minimum corner of the box" or "Minimum point".

File: MonoGame.Framework/BoundingBox.cs [hunk 3]
```diff
@@ -29,6 +41,11 @@ public struct BoundingBox : IEquatable<BoundingBox>
 
         #region Public Constructors
 
+        /// <summary>
+        /// Creates an axis aligned bounding box defining minimum and maximum edge.
+        /// </summary>
+        /// <param name="min">Minimum edge.</param>
```
Comment (mgarstenauer) [inline MonoGame.Framework/BoundingBox.cs]: Instead of "edge" I would write "corner", "point" or "position".

**Comment by mgarstenauer** [comment]
Hi @trommlbomml, can you describe a use case where this new method is required. IMHO, this functionality fits better in a third-party library as an extension method.

**Comment by trommlbomml** [comment]
> can you describe a use case where this new method is required.
Yes. When it comes to collision in player movement, you may face the problem you want to slide along the wall. You can implement it on your own by testing each axis or using the universal function I provided. You just need to move your player, update bounding box, call the SAT function and move your character back the vector you get when there is any intersection. This is very handy, especially when you want to move precisely along your colliders.

File: MonoGame.Framework/BoundingBox.cs (hunk 3 above)
Comment (trommlbomml) [inline MonoGame.Framework/BoundingBox.cs]: Yes, you are right. I dunno if @tomspilman finds it good to fix this when there is already a working pull request state.

**Comment by mgarstenauer** [comment]
> but this problem is so common ...

I am not convinced that this is a common use case. It is of very limited use. It only works with 90° axis-aligned walls. A character is usually not represented as an AABB. It is usually a capsule or sphere to faciliate stepping and smooth sliding around edges.

**Comment by trommlbomml** [comment]
The current implementation is very specific to this case, you are right. The technique is not specifically restricted to AABB, just to convex shapes, which really makes sense as you often have to split concave objects for other checks. You could easily extend this method to all existing shapes in MonoGame. 

When this is your reason I can accept this. The system could be extended to be used with all existing 3D shapes in MonoGame (and 2D).

**Comment by mgarstenauer** [comment]
Yes, that is my reason. The current implementation it is too limited. I don't know about an extended implementation. I would need to see the API. (I have been down this road [before](https://www.digitalrune.com/Documentation/html/335dc86a-c68d-4d7b-8641-81dd80de5e76.htm). Collision detection needs to be well thought-out. Difficult topics like this can be solved in a third-party library. Or perhaps we should start an incubation project for testing new functionality before adding it to the public MonoGame API? - Just an idea.)

But: I am not a decision maker. Of course, others can weigh in as well.

**Comment by trommlbomml** [comment]
I get your points. I think this is not a specific API but an extension to collision detection with more detailed information. the API would be the same ad already implemented. A collision test returns instead of bool or intersection Kind this separation information. so there will be something like boundingSphere.IntersectSat(myObb). For each possible combination there will be a method in this Signature. you can decide to use the methods with vor without separation.

**Comment by KonajuGames** [comment]
We have been careful to not try to make MonoGame a Swiss army knife of game
development.  It is a low level framework for the creation of many types of
games.  Functionality such as collision detection we believe should live in
external libraries so the developer can choose or implement the system that
suits their game best, out none at all if nothing is required.

The existing intersect routines and bounding types in MonoGame are here
because not only were they in XNA, but also because they are commonly used
in frustum culling and UI hit testing.

This PR is the kind of functionality that would be ideal in an external
library, along with a suite of other collision functions.

**Comment by tomspilman** [comment]
As much as i hate it...  someone has to be the bad guy here.

I agree with @KonajuGames and @mgarstenauer that this likely belongs in a 3rd-party library.  I suggest talking to @craftworkgames about his [MonoGame.Extended](https://github.com/craftworkgames/MonoGame.Extended) project.

In the future we will make these calls quickly.   If it isn't clear in a couple of days if a PR is good or not... i will close it.  It doesn't help anyone involved for us to be wishy washy on a PR and it stay open for a year.

**Comment by craftworkgames** [comment]
Yep. Thanks @tomspilman. I believe this could be a good fit for our library. We already have a number of extensions on the standard shapes.

**Comment by trommlbomml** [comment]
This is a good So I recreate the changes in the MonoGame.Extended repository and I accept to reject my pull request.

**Pull request closed without merging**